bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/OpenhabGraalJSScriptEngine.java

```diff
@@ -227,11 +232,20 @@ protected void beforeInvocation() {
     }
 
     @Override
-    public Object invokeFunction(String s, Object... objects) throws ScriptException, NoSuchMethodException {
-        // Synchronize multi-thread access to avoid exceptions when reloading a script file while the script is running
-        synchronized (lock) {
-            return super.invokeFunction(s, objects);
-        }
+    protected void afterInvocation() {
+        lock.unlock();
+    }
+
+    @Override
+    protected Object afterInvocation(Object obj) {
+        afterInvocation();
+        return obj;
+    }
```

Comment:
Are you sure that the overridden methods are always the same as they are now (i.e. do nothing/return object)? Otherwise it would be good to do something like

```suggestion
    protected void afterInvocation() {
        lock.unlock();
        super.afterInvocation();
    }

    @Override
    protected Object afterInvocation(Object obj) {
        lock.unlock();
        return super.afterInvocation(obj);
    }
```

Reply:
Note that the methods that are overriden here are newly introduced together with their overrides. I followed the pattern of `beforeInvocation` when I added `afterInvocation`: have a method at the `InvocationInterceptingXXX` class and override it in the `OpenhabGraalJSScriptEngine`.

Reply:
The `beforeInvocation` override also doesn't call it's super method, so I'd keep as it is and resolve this comment.